File: frontend/src/core/crypto/recoveryKey.ts
const ALPHABET = '0123456789ABCDEFGHJKMNPQRSTVWXYZ'

const DECODE: Record<string, number> = {}
for (let i = 0; i < ALPHABET.length; i++) {
  DECODE[ALPHABET[i]] = i
}
// Crockford leniency: treat visually-ambiguous characters as canonical equivalents
DECODE['O'] = DECODE['0']
DECODE['I'] = DECODE['1']
DECODE['L'] = DECODE['1']
for (const key of Object.keys(DECODE)) {
  DECODE[key.toLowerCase()] = DECODE[key]
}

export const RECOVERY_KEY_RAW_BYTES = 20
export const RECOVERY_KEY_DISPLAY_LENGTH = 39   // 32 alphabet chars + 7 hyphens

export class InvalidRecoveryKeyError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'InvalidRecoveryKeyError'
  }
}

/** Generate a fresh 160-bit recovery key in display form. */
export function generateRecoveryKey(): string {
  const raw = new Uint8Array(RECOVERY_KEY_RAW_BYTES)
  crypto.getRandomValues(raw)
  return encodeRecoveryKey(raw)
}

/** Encode 20 raw bytes as `XXXX-XXXX-XXXX-XXXX-XXXX-XXXX-XXXX-XXXX` (39 chars). */
export function encodeRecoveryKey(raw: Uint8Array): string {
  if (raw.length !== RECOVERY_KEY_RAW_BYTES) {
    throw new InvalidRecoveryKeyError(`raw must be ${RECOVERY_KEY_RAW_BYTES} bytes, got ${raw.length}`)
  }
  // Big-endian int, then repeatedly divmod 32.
  let n = 0n
  for (const b of raw) n = (n << 8n) | BigInt(b)
  const chars: string[] = []
  for (let i = 0; i < 32; i++) {
    chars.push(ALPHABET[Number(n & 31n)])
    n >>= 5n
  }
  const compact = chars.reverse().join('')
  return [
    compact.slice(0, 4), compact.slice(4, 8), compact.slice(8, 12), compact.slice(12, 16),
    compact.slice(16, 20), compact.slice(20, 24), compact.slice(24, 28), compact.slice(28, 32),
  ].join('-')
}

/** Strip whitespace + hyphens, uppercase. Result is 32 significant chars on success. */
export function normaliseRecoveryKey(input: string): string {
  return input.replace(/[\s-]/g, '').toUpperCase()
}

/** Decode a display-form or compact-form recovery key to its 20 raw bytes. */
export function decodeRecoveryKey(input: string): Uint8Array {
  const s = normaliseRecoveryKey(input)
  if (s.length !== 32) {
    throw new InvalidRecoveryKeyError(`must have 32 significant characters, got ${s.length}`)
  }
  let n = 0n
  for (const ch of s) {
    const v = DECODE[ch]
    if (v === undefined) throw new InvalidRecoveryKeyError(`invalid character: ${ch}`)
    n = n * 32n + BigInt(v)
  }
  const out = new Uint8Array(RECOVERY_KEY_RAW_BYTES)
  for (let i = RECOVERY_KEY_RAW_BYTES - 1; i >= 0; i--) {
    out[i] = Number(n & 0xffn)
    n >>= 8n
  }
  return out
}

File: backend/modules/user/_recovery_key.py
"""Recovery-key generation, formatting, and parsing (Crockford-Base32).

Key format for display and entry: ``XXXX-XXXX-XXXX-XXXX-XXXX-XXXX-XXXX-XXXX``
(8 groups of 4 characters, hyphens for readability). The raw underlying
material is 20 bytes (160 bit), encoded as 32 Crockford-Base32 characters.
Crockford is chosen over standard Base32 because it excludes the visually
ambiguous characters I, L, O, U and accepts ``O``/``I``/``L`` on input as
synonyms for ``0``/``1``/``1`` — friendlier when a user copies the key
from paper or a phone display.
"""

from __future__ import annotations

import os

RECOVERY_KEY_RAW_BYTES = 20
RECOVERY_KEY_DISPLAY_LENGTH = 32 + 7  # 32 chars + 7 hyphens = 39

_ALPHABET = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"  # Crockford-Base32, no I L O U
_DECODE: dict[str, int] = {c: i for i, c in enumerate(_ALPHABET)}
# Crockford input leniency
for src, dst in {"O": "0", "I": "1", "L": "1"}.items():
    _DECODE[src] = _DECODE[dst]
    _DECODE[src.lower()] = _DECODE[dst]
for c, i in list(_DECODE.items()):
    _DECODE[c.lower()] = i


class InvalidRecoveryKeyError(ValueError):
    """Raised when a recovery key fails to decode."""


def generate_recovery_key() -> str:
    """Generate a freshly random recovery key in display form."""
    raw = os.urandom(RECOVERY_KEY_RAW_BYTES)
    return encode_recovery_key(raw)


def encode_recovery_key(raw: bytes) -> str:
    """Encode 20 bytes as the 39-character display string with hyphens."""
    if len(raw) != RECOVERY_KEY_RAW_BYTES:
        raise InvalidRecoveryKeyError(
            f"raw must be {RECOVERY_KEY_RAW_BYTES} bytes, got {len(raw)}"
        )
    # Convert bytes to a big-endian integer, then repeatedly divmod by 32.
    n = int.from_bytes(raw, "big")
    chars: list[str] = []
    for _ in range(32):
        n, rem = divmod(n, 32)
        chars.append(_ALPHABET[rem])
    encoded = "".join(reversed(chars))
    return "-".join(encoded[i : i + 4] for i in range(0, 32, 4))


def normalise_recovery_key(user_input: str) -> str:
    """Strip whitespace, remove hyphens, uppercase. Result is 32 chars on success."""
    stripped = "".join(ch for ch in user_input if not ch.isspace() and ch != "-")
    return stripped.upper()


def decode_recovery_key(user_input: str) -> bytes:
    """Parse a user-supplied recovery key to its 20 raw bytes.

    Accepts the display form (with hyphens) or the compact form (no hyphens),
    case-insensitive, with Crockford leniency on O/I/L.
    """
    normalised = normalise_recovery_key(user_input)
    if len(normalised) != 32:
        raise InvalidRecoveryKeyError(
            f"recovery key must have 32 significant characters, got {len(normalised)}"
        )
    n = 0
    for ch in normalised:
        try:
            n = n * 32 + _DECODE[ch]
        except KeyError as exc:
            raise InvalidRecoveryKeyError(f"invalid character: {ch!r}") from exc
    return n.to_bytes(RECOVERY_KEY_RAW_BYTES, "big")
